### PR TITLE
add bycolumn function

### PR DESCRIPTION
**Pull request by simonbyrne**
This should make it easier to apply multiple column operations in a single pass, and leverage threading.

Example:
```
bycolumn(axes(f)) do colidx
    @. dg[colidx] = ∇(f[colidx])
    @. df[colidx] = div(g[colidx])
end
```

The reason to use `getindex` here is that it doesn't get broadcasted by `@.`, so makes writing these operations easier.

- [ ] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [ ] Unit tests are included OR N/A.
- [ ] Code is exercised in an integration test OR N/A.
- [ ] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Review by charleskawczynski — COMMENTED** [review]
I'll try testing this out with TC on Monday

**Comment by charleskawczynski** [comment]
Better yet, can you add a test that tests this out with ntuple fields? ( and does the overloading/indexing like we do in TC? )

Basically, in a test, you want to test this, with ntuple fields alongside what's done in #515.

File: src/Fields/indices.jl [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+# Example
+
+```julia
+∇ = GradientF2C()
+div = DivergenceC2F()
+
+bycolumn(axes(f)) do colidx
+    @. dg[colidx] = ∇(f[colidx])
+    @. df[colidx] = div(g[colidx])
```
Comment (charleskawczynski) [inline src/Fields/indices.jl:47]:
```suggestion
    @. ∇f[colidx] = ∇(f[colidx])
    @. df[colidx] = div(∇f[colidx])
```

**Comment by charleskawczynski** [comment]
Actually, I think that this should be fine since we're overloading `getindex` on `ColumnIndex` (`Base.getindex(field::Field, colidx::ColumnIndex) = column(field, colidx)`).

File: src/Fields/indices.jl [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+bycolumn(axes(f)) do colidx
+    @. ∇f[colidx] = ∇(f[colidx])
+    @. df[colidx] = div(∇f[colidx])
+end
+```
+"""
+function bycolumn(fn, space::Spaces.SpectralElementSpace1D)
+    Nh = Topologies.nlocalelems(space)
+    Nq = Spaces.Quadratures.degrees_of_freedom(Spaces.quadrature_style(space))
+    Threads.@threads for h in 1:Nh
```
Comment (charleskawczynski) [inline src/Fields/indices.jl:54]: Do we need `@inbounds`? And the `@threads` is ignored when no threading is used, correct?

File: src/Fields/indices.jl [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            fn(ColumnIndex((i,), h))
+        end
+    end
+    return nothing
+end
+function bycolumn(fn, space::Spaces.SpectralElementSpace2D)
+    Nh = Topologies.nlocalelems(space)
+    Nq = Spaces.Quadratures.degrees_of_freedom(Spaces.quadrature_style(space))
+    Threads.@threads for h in 1:Nh
+        for j in 1:Nq, i in 1:Nq
```
Comment (charleskawczynski) [inline src/Fields/indices.jl:65]: Do we need `@inbounds`? And the `@threads` is ignored when no threading is used, correct?

Comment (simonbyrne) [reply src/Fields/indices.jl:65]: I'm not sure: I think `@threads` still has some overhead, so we may want to make it optional.

**Pull request merged**